python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -1,45 +1,182 @@
-from typing import Callable
+import logging
+from typing import Callable, Iterator, TypeVar, List, Any
 
+import ray
+from ray.exceptions import RayActorError
 from ray.util.sgd.v2.worker_group import WorkerGroup
 
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_next():
+    """Fetch next item from result queue on each worker."""
+    # TODO: Implement along with sgd.report()
+    return None
+
+
+def fetch_all():
+    """Fetch all items from result queue on each worker."""
+    # TODO: Implement along with sgd.report()
+    return [None]
+
 
 class BackendConfig:
+    """Parent class for configurations of backends (torch, horovod, etc.)"""
+
     @property
     def backend_name(self):
         raise NotImplementedError
 
     def validate(self, name):
-        assert name == self.backend_name
+        assert name == self.backend_name(), (name, self.backend_name)
 
 
 class BackendExecutor:
+    """Main execution class for SGD backends (torch, tensorflow, etc.).
+
+    This class holds a worker group and is responsible for executing the
+    training function on the workers, and collecting intermediate results
+    from ``sgd.report()``.
+
+    Args:
+        backend_config (BackendConfig): The configurations for this
+            specific backend.
+        num_workers (int): Number of workers to use for training.
+        num_cpus_per_worker (float): Number of CPUs to use per worker.
+        num_gpus_per_worker (float): Number of GPUs to use per worker.
+    """
+
     def __init__(self,
                  backend_config: BackendConfig,
                  num_workers: int = 1,
-                 num_cpus_per_worker: int = 1,
-                 num_gpus_per_worker: int = 0):
+                 num_cpus_per_worker: float = 1,
+                 num_gpus_per_worker: float = 0):
         self._backend_config = backend_config
         self._num_workers = num_workers
         self._num_cpus_per_worker = num_cpus_per_worker
         self._num_gpus_per_worker = num_gpus_per_worker
 
+        self.worker_group = DeactivatedWorkerGroup()
+
     def start(self):
+        """Starts the worker group."""
         self.worker_group = WorkerGroup(self._num_workers,
                                         self._num_cpus_per_worker,
                                         self._num_gpus_per_worker)
 
-    def execute(self, train_func: Callable):
-        pass
+    def execute(self, train_func: Callable[[], T]) -> Iterator[Any]:
+        """Executes training function on all workers and yield results.
+
+        The provided function must have any required arguments.
+
+        Args:
+            train_func (Callable): The training function to run on each
+                worker. It must not have any required arguments.
```

Comment:
The doc for `run` mentions `It must accept ``config`` as an argument.` - we can synchronize these and update them when the `config` use case is implemented (e.g. Tune).